src/gui/guiutility.cpp

```diff
@@ -38,47 +45,76 @@ Q_LOGGING_CATEGORY(lcUtility, "gui.utility", QtInfoMsg)
 namespace {
 
 #ifdef Q_OS_WIN
-void watchWM()
+// TODO: 2.11 move to the new Platform class
+struct
+{
+    HANDLE windowMessageWatcherEvent = CreateEventW(nullptr, true, false, L"watchWMEvent");
+    bool windowMessageWatcherRun = true;
+} watchWMCtx;
+
+void startShutdownWatcher()
 {
     // Qt only receives window message if a window was displayed at least once
     // create an invisible window to handle WM_ENDSESSION
-    QThread::create([] {
+    // We also block a system shutdown until we are properly shutdown our selfs
+    // In the unlikely case that we require more than 5s Windows will require a fullscreen message
+    // with our icon, title and the reason why we are blocking the shutdown.
+    new std::thread([] {
         WNDCLASS wc = {};
         wc.hInstance = GetModuleHandle(nullptr);
         wc.lpszClassName = L"ocWindowMessageWatcher";
+#if MIRALL_VERSION_MINOR > 9
+// TODO: for now we won't display a proper icon
+#error "Please add the QtWinExras dependency"
+        wc.hIcon = QtWin::toHICON(Theme::instance()->applicationIcon().pixmap(64, 64));
+#endif
         wc.lpfnWndProc = [](HWND hwnd, UINT msg, WPARAM wParam, LPARAM lParam) -> LRESULT {
             //            qDebug() << MSG { hwnd, msg, wParam, lParam, 0, {} };
             if (msg == WM_QUERYENDSESSION) {
+                qCDebug(OCC::lcUtility) << "Received WM_QUERYENDSESSION";
                 return 1;
             } else if (msg == WM_ENDSESSION) {
-                qCInfo(OCC::lcUtility) << "Received WM_ENDSESSION quitting";
-                QTimer::singleShot(0, qApp, &QApplication::quit);
+                qCDebug(OCC::lcUtility) << "Received WM_ENDSESSION quitting";
+                QMetaObject::invokeMethod(qApp, &QApplication::quit);
+                QElapsedTimer shutdownTimer;
+                if (lParam == ENDSESSION_LOGOFF) {
+                    // block the windows shutdown until we are done
+                    const QString description = QApplication::translate("Utility", "Shutting down %1").arg(Theme::instance()->appNameGUI());
+                    OC_ASSERT(ShutdownBlockReasonCreate(hwnd, reinterpret_cast<const wchar_t *>(description.utf16())));
```

Comment:
The Windows docs recommend to provide a short, reasonable explanation why the shutdown is postponed. For instance, "Closing SQL database properly".

Reply:
Way to technical